proposals/experiment-tracking-proposal.md

```diff
@@ -0,0 +1,84 @@
+
+## Example user stories
+* I'm a data scientist working on a problem. I'm looking for easy way to compare multiple training jobs with multiple sets of hyperparameters. I would like to be able to select top 5 jobs measured with P1 score and examine which model architecture, hyperparameters, dataset and initial state contributed to this score. I would want to compare these 5 together in highly detailed way (for example via tensorboard). I would like rich UI to navigate models without need to interact with infrastructure.
+* I'm part of big ML team in company. Our whole team works on single problem (for example search) and every person builds their models. I'd like to be able to compare my models with others. I want to be safe that nobody will accidentally delete model I'm working on.
+* I'm cloud operator in ML team. I would like to take current production model (architecture+hyperparams+training state) and retrain it with new data as it becomes available. I would want to run suite of tests and determine if new model performs better. If it does, I'd like to spawn tf-serving (or seldon) cluster and perform rolling upgrade to new model.
+* I'm part of highly sophisticated ML team. I'd like to automate retraining->testing->rollout for models so they can be upgraded nightly without supervision.
+* I’m part of the ML team in company. I would like to be able to track my parameters used to train an ML job and track the metrics produced. I would like to have isolation so I can find the experiments I worked on. I would like the ability to compare multiple training rules side by side and pick the best model. Once I select the best model I would like to deploy that model to production. I may also want to test my model with historical data to see how well it performs and maybe roll out my experiment to a subset of users before fully rolling out to all users.
+
+
+## Scale considerations
+
+We need to support scale in tens of thousands of models. Potentially adding garbage collection above this. Within this scale we need to be able to quickly select best models
+for particular problem.
+
+## Model provenance
+
+Another feature commonly asked for is model provenance. It's crucial to be able to reproduce results. For every model we need to record:
+
+* Inital state, whether it's random weights or based on preexisting models
+* Dataset used for training
+* Dataset used for testing
+* Feature engineering pipeline used
+* Katib study id
+* Model architecture (code used)
+* Hyperparameters
```

Comment:
If the code to create the model is in a VCS, e.g. git, it should also track the version of the code used to create the model

Reply:
agree, that's what I meant by "model architecture". But good idea would be to make it point to
* code (including commit id)
* docker image